pkg/ui/server.go

```diff
@@ -0,0 +1,282 @@
+		successDetails := &SuccessDetails{
+			PageTitle:    "JVS - Successful token retrieval",
+			Token:        token,
+			TargetOrigin: formDetails.Origin,
+			WindowName:   formDetails.WindowName,
+		}
+
+		render(w, templates["success"], successDetails)
+	}
+}
+
+// Checks the origin parameter against all entries in the allow list.
+func validateOrigin(originParam string, allowList []string) bool {
+	// origin: shop.acme.com
+	// allowList: [acme.com,*]
+
+	// Check if origin is localhost or private ip
+	if strings.HasPrefix(originParam, "http://localhost") {
+		return true
+	}
+
+	originSplit := strings.Split(originParam, ".")
+
+	for _, domain := range allowList {
+		// special case, allow everything
+		if domain == "*" {
+			return true
+		}
+
+		domainSplit := strings.Split(domain, ".")
+
+		// this domain is longer than the origin, skip over it
+		if len(domainSplit) > len(originSplit) {
+			continue
+		}
+
+		// compare the origin and current domain from right to left
+		for i, j := len(domainSplit)-1, len(originSplit)-1; i >= 0 && j >= 0; i, j = i-1, j-1 {
+			// not a wildcard reference and no match, proceed to the next domain candidate in the allow list
+			if domainSplit[i] != "*" && domainSplit[i] != originSplit[j] {
+				break
+			}
+
+			if i == 0 {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func validateForm(formDetails *FormDetails) bool {
+	formDetails.Errors = make(map[string]string)
+
+	if !isValidOneOf(formDetails.Category, categories) {
+		formDetails.Errors["Category"] = "Category must be selected"
+	}
+
+	if strings.TrimSpace(formDetails.Reason) == "" {
+		formDetails.Errors["Reason"] = "Reason is required"
+	}
+
+	if !isValidOneOf(formDetails.TTL, ttls) {
+		formDetails.Errors["TTL"] = "TTL is required"
+	}
+
+	return len(formDetails.Errors) == 0
+}
+
+func isValidOneOf(selection string, options []string) bool {
+	for _, v := range options {
+		if v == selection {
+			return true
+		}
+	}
+	return false
+}
+
+func getFormDetails(r *http.Request) *FormDetails {
+	return &FormDetails{
+		WindowName: r.FormValue("windowname"),
+		Origin:     r.FormValue("origin"),
+		Category:   r.FormValue("category"),
+		Reason:     r.FormValue("reason"),
+		TTL:        r.FormValue("ttl"),
+		PageTitle:  "JVS - Justification Request System",
+		Content: Content{
+			UserLabel:     "User",
+			CategoryLabel: "Category",
+			ReasonLabel:   "Reason",
+			TTLLabel:      "TTL",
+			Categories: []Pair{
+				{
+					Key:  "explanation",
+					Text: "Explanation",
+				},
+				{
+					Key:  "breakglass",
+					Text: "Breakglass",
+				},
+			},
+			TTLs: []Pair{
+				{
+					Key:  "15",
+					Text: "15m",
+				},
+				{
+					Key:  "30",
+					Text: "30m",
+				},
+				{
+					Key:  "60",
+					Text: "1h",
+				},
+				{
+					Key:  "120",
+					Text: "2h",
+				},
+				{
+					Key:  "240",
+					Text: "4h",
+				},
+			},
+		},
+	}
+}
+
+func render(w http.ResponseWriter, tmpl *template.Template, data any) {
```

Comment:
This is quite dangerous.

Reply:
Curious to learn more

Reply:
Is there an action I can take to make it not dangerous?

Reply:
There are a few problems here:

1. You should accept an `io.Writer`, not an `http.ResponseWriter`. That will make testing easier (and `ResponseWriter` is a subset of `Writer`).

2. Rendering directly to the writer. HTTP is chunked - if a template is sufficiently large such that it exceeds the template buffer, the partial results will be written to the writer, flushed, and then rendering will continue. If later, while processing the template, an error occurs (even a syntax error, since templates aren't "compiled"), the template parsing will fail and return an error. Except, you've already sent part of a response with an implicit 200 OK. So you can't return an error to the client anymore - you just serve half content which is a pretty bad experience. For this reason, you usually render into a buffer and then iff the buffer renders do you flush the buffer to the writer.

3. Once you switch to a buffer, there's a memory management problem. You don't want to allocate a new buffer for each page load; memory will spike and GCs will be very expensive. It's also possible to DOS a service like this. Instead, you need to use a buffer pool to save on allocations and GC sweeps.

See EN for an example.

Reply:
Got it, i spent some time understanding the general setup for EN. Working on making those upgrades. Will update later and re-request review when its ready.